lib/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,272 @@
+<?php
+/**
+ * Block template functions.
+ *
+ * @package Gutenberg
+ * @since 6.7.0
+ */
+
+if ( ! class_exists( 'WP_Block_Templates_Registry' ) ) {
+	/**
+	 * Core class used for interacting with block templates.
+	 *
+	 * @since 6.7.0
+	 */
+	final class WP_Block_Templates_Registry {
+		/**
+		 * Registered block templates, as `$name => $instance` pairs.
+		 *
+		 * @since 6.7.0
+		 * @var WP_Block_Template[] $registered_block_templates Registered block templates.
+		 */
+		private $registered_block_templates = array();
+
+		/**
+		 * Container for the main instance of the class.
+		 *
+		 * @since 6.7.0
+		 * @var WP_Block_Templates_Registry|null
+		 */
+		private static $instance = null;
+
+		/**
+		 * Registers a block template.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string|WP_Block_Template $template_name Block template name including namespace, or alternatively
+		 *                                                a complete WP_Block_Template instance. In case a WP_Block_Template
+		 *                                                is provided, the $args parameter will be ignored.
+		 * @param array                    $args          Optional. Array of block template arguments.
+		 * @return WP_Block_Template|false The registered block template on success, or false on failure.
+		 */
+		public function register( $template_name, $args = array() ) {
```

Comment:
I'm still not sure why to provide two ways of passing the template and not just one. I get it that it's like that for block type registry, but in other registries isn't and I don't see any reason to not keep things simpler.

Reply:
Just to confirm I understand, you mean that instead of having `gutenberg_register_template()` and `WP_Templates_Registry::register()` we should only have `gutenberg_register_template()`?

Reply:
No, I meant why have a way to provide a `string` name and `args`, we also support `template_name` to be an object. I think it's better to provide a single way of registering and I think in other registries we have `string` name and `args`.

Reply:
Oh, it makes sense. I think it was a carry over from the blocks registry. Changed in ab18e5a47926dd31ec3ca96bc233fac0da1b1dfa.